### PR TITLE
Papia creates myteam filter

### PR DESCRIPTION
**Pull request by papiasharmin**
# Description
(PRIORITY High) Jae: Create “My Team” filter for Tasks Tab and Leaderboard (WIP Papia)
On the Dashboard, we need a “My Team” button and filter. Clicking this button should filter both the Tasks Tab and Leaderboard to show only the people on that person’s team. Then they should be able to use the 24h, 48h, etc. buttons and only see their team members. 
Make it the Default option for Core Team members
This is for Owner, Admin and Core Team classes to have a way to only see teams they are on in both the Leaderboard and Tasks Tab views (see below)
This is needed because these classes see ALL users and that makes useless all the tools for easily reviewing the work of just their team 

## Main changes explained:
Added 'My Team' toggle button. Added a toggleTeamView function and getMyTeam helper function to get data for logged in user's team members. Remove fetching leaderBoardData from mapStateToProps to leaderboard component, as it is fetching the wrong state.

## How to test:
1.check into current branch
2.do npm run start:local to run this PR locally
3.Log as admin or owner user or core team,On dashboard user will show team toggler button and their team members by default if they are assigned to any team, on leadderboard and task tab. If they are not assigned to any team, will not see the button and will see all members.
toggele the button to see if it shows all members and the person's team members alternatively. And also able to ftech correct data for clicking 24hr,48hr,72hr buttons. Also check the responsivenes.
4.Log in as volunteer or manager to see they can't see the filter button, and correct data of members are fetched on task and leaderboard 

## Screenshots or videos of changes:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/86232710/9b0135b4-5567-4b28-8839-413cb95313f6

### CODE REVIEW

**Comment by AnishPandita** [comment]
Hi @papiasharmin I tested the PR and your suggested changes seem to work. The team toggle button and assigned functionality works as intended.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/111321945/98acebd8-9c38-4d48-85e1-af372d0db111

However when i logged in as volunteer account my dashboard seemed frozen and I was not able to click anything. Could someone else please confirm if thats the case or something might be wrong at by browser end.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/111321945/d90b46be-8f4c-46f5-a39d-c36f0b171f18

File: src/components/Dashboard/Dashboard.jsx [hunk 1]
```diff
@@ -31,12 +34,24 @@ export const Dashboard = props => {
...
   }, []);
 
-  useEffect(() => {
-    if (props.match.params && props.match.params.userId && userId != props.match.params.userId) {
-      userId = props.match.params.userId;
-      getUserProfile(userId);
-    }
-  }, [props.match]);
+  useEffect(()=>{
+  },[userProfile])
```
Comment (stevenpaalz) [inline src/components/Dashboard/Dashboard.jsx:38]: What is the purpose of this useEffect hook here?

**Review by stevenpaalz — DISMISSED** [review]
I left one comment on code for you to address, but other than that this appears to be functioning as intended.

On admin account:
https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/121642083/837a6a4f-296e-4934-bea5-5fd646385236

On volunteer/manager account:
Screenshot 2023-06-15 at 10 33 39 PM

**Comment by yufeiStella** [comment]
Hi @papiasharmin I test your function and it works perfectly!
https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/111897624/37374201-e1f2-46e4-9a5b-b63fce4e775c

**Review by bienzguoa — APPROVED** [review]
Works perfect!

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/92126559/15b41637-5af0-4889-98a9-acca19775239

**Comment by papiasharmin** [comment]
> Hi @papiasharmin I tested the PR and your suggested changes seem to work. The team toggle button and assigned functionality works as intended.
> 
>  Screen.Recording.2023-06-15.at.2.24.24.PM.mov 
> However when i logged in as volunteer account my dashboard seemed frozen and I was not able to click anything. Could someone else please confirm if thats the case or something might be wrong at by browser end.
> 
>  Screen.Recording.2023-06-15.at.2.34.35.PM.mov

Hi, @AnishPandita , I have also seen that issue, this is happening if you were in admin page where  members data load take some time, if you logout before the data load finishes, the page become still, until you reload the page.  I think there should be a way to abort fetching if you log out.

**Review by Chuehleo — DISMISSED** [review]
Everthing work as intended, great job!

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/122568562/c782284c-3749-4e92-8081-8893b46dca21

File: src/components/Dashboard/Dashboard.jsx (hunk 1 above)
Comment (papiasharmin) [reply src/components/Dashboard/Dashboard.jsx:38]: Hello, @stevenpaalz, thank you, I forgot to remove that.

**Comment by cvtqx** [comment]
@papiasharmin I am testing this PR, but I only see the button in the Leaderboard but not in the Tasks:
Screen Shot 2023-06-16 at 11 30 51 AM

The toggle button works as intended in the leaderboard switching the view between 'my teams' and 'view all'.

Perhaps I am missing something when I create a team which is why I am not seeing anything under Tasks. Do you have any idea of what the issue might be?

**Review by eduhorta — DISMISSED** [review]
Hi, @papiasharmin!

The feature is working as intended! Great Job!! 👍

Volunteer account:
Screenshot 2023-06-16 15-42-33

Admin account:
Screenshot 2023-06-16 15-43-39

**Comment by papiasharmin** [comment]
> @papiasharmin I am testing this PR, but I only see the button in the Leaderboard but not in the Tasks: Screen Shot 2023-06-16 at 11 30 51 AM
> 
> The toggle button works as intended in the leaderboard switching the view between 'my teams' and 'view all'.
> 
> Perhaps I am missing something when I create a team which is why I am not seeing anything under Tasks. Do you have any idea of what the issue might be?

Hello, @cvtqx, if you have just assigned a team that might be a problem. I have checked the functionality several time, I will test it again and let you know.

**Review by OAkinbode — COMMENTED** [review]
Hello Papia, 

The PR works wonderfully. Nice work.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/104801652/32a3a49b-e6b0-4a42-a6fc-9665800892aa

**Comment by papiasharmin** [comment]
> @papiasharmin I am testing this PR, but I only see the button in the Leaderboard but not in the Tasks: Screen Shot 2023-06-16 at 11 30 51 AM
> 
> The toggle button works as intended in the leaderboard switching the view between 'my teams' and 'view all'.
> 
> Perhaps I am missing something when I create a team which is why I am not seeing anything under Tasks. Do you have any idea of what the issue might be?

Hello @cvtqx, can you please rerun the local server then login to your account. I believe you will see the button. User's team array not updating after just assigning team unless you rerun the server. I have console logged the user profile to ensure that. That could be a bug.

File: src/components/LeaderBoard/Leaderboard.jsx [hunk 2]
```diff
@@ -24,37 +32,140 @@ function useDeepEffect(effectFunc, deps) {
...
+    if (loggedInUser.role !== 'Administrator' && loggedInUser.role !== 'Owner' && loggedInUser.role !== 'Core Team') {
+      leaderBoardData = leaderBoardData.filter(element => {
+        return element.weeklycommittedHours > 0 || loggedInUser.userid === element.personId;
+      });
+    }
+  
+    if (leaderBoardData.length) {
+      let maxTotal = maxBy(leaderBoardData, 'totaltime_hrs').totaltime_hrs || 10;
+      leaderBoardData = leaderBoardData.map(element => {
+        element.didMeetWeeklyCommitment =
```
Comment (cone-7) [inline src/components/LeaderBoard/Leaderboard.jsx:75]:
```suggestion
        element.didMeetWeeklyCommitment = element.totaltangibletime_hrs >= element.weeklycommittedHours;
```

File: src/components/LeaderBoard/Leaderboard.jsx [hunk 3]
```diff
@@ -24,37 +32,140 @@ function useDeepEffect(effectFunc, deps) {
...
+        if(member.length > 0){
+          member.forEach(user => {
+            res = res.filter(item => user._id !== item._id);
+          });
+        }
+        member = member.concat(res);
+    }
+    const filteredlist = leaderboarddata.filter(user => {
+      for(let i = 0; i < member.length; i++){
+        if(user.personId == member[i]._id){
```
Comment (cone-7) [inline src/components/LeaderBoard/Leaderboard.jsx:143]: We should use `'==='`;

File: src/components/LeaderBoard/Leaderboard.jsx [hunk 4]
```diff
@@ -160,6 +272,14 @@ const LeaderBoard = ({
           }}
         />
       </h3>
+      {(loggedInUser.role === 'Owner' || loggedInUser.role === 'Administrator' || loggedInUser.role == 'Core Team') && userTeams?.length > 0 && <button className='circle-border my-team' style={{
```
Comment (cone-7) [inline src/components/LeaderBoard/Leaderboard.jsx:275]: we are doing this validation in different parts of the file, maybe it is worth to create a function to check this condition (if there is a utils file where this function can be added better) and create a variable to use where it is required.

File: src/components/TeamMemberTasks/TeamMemberTasks.jsx [hunk 5]
```diff
@@ -87,34 +89,64 @@ const TeamMemberTasks = props => {
...
+    if((props.userProfile.role=== 'Owner' || props.userProfile.role === 'Administrator' || props.userProfile.role === 'Core Team') && props.userProfile.teams.length > 0){
+      setisTeamTab(true)
+      setisLoadingmember(true)
+    }
+  },[props.userProfile])
+
+  //which data to show depending on user role
+  useEffect(()=>{
+    if(userRole){
+      if(userRole === 'Owner' || userRole === 'Administrator' || userRole === 'Core Team' && props.userProfile.teams.length > 0){
```
Comment (cone-7) [inline src/components/TeamMemberTasks/TeamMemberTasks.jsx:104]: Same condition as above.

**Comment by cvtqx** [comment]
> > @papiasharmin I am testing this PR, but I only see the button in the Leaderboard but not in the Tasks: Screen Shot 2023-06-16 at 11 30 51 AM
> > The toggle button works as intended in the leaderboard switching the view between 'my teams' and 'view all'.
> > Perhaps I am missing something when I create a team which is why I am not seeing anything under Tasks. Do you have any idea of what the issue might be?
> 
> Hello @cvtqx, can you please rerun the local server then login to your account. I believe you will see the button. User's team array not updating after just assigning team unless you rerun the server. I have console logged the user profile to ensure that. That could be a bug.

@papiasharmin I've re-tested this PR and both buttons are displayed properly, toggle functionality also works well! When logged in as volunteer or manager the buttons are not displayed.

Admin view:
Screen Shot 2023-06-20 at 3 15 37 PM

**Review by jonathancarpena — APPROVED** [review]
@papiasharmin Working as intended, great job! 👍

**Admin View:**



**Volunteer View:**

**Comment by beblicarl** [comment]
This works as intended. Awesome !!

[screencast-localhost_3000-2023.06.29-09_12_19.webm](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/68056104/cc059445-8e3e-4f53-a5cd-2f3ec14001a0)

**Review by beblicarl — DISMISSED** [review]

**Review by mahamakhan — COMMENTED** [review]
Spent a lot of time figuring it out but I only see the view all button which is not visible on volunteer account so that works but I can't see others. Maybe I am doing something wrong. 


Screenshot 2023-06-30 203509
Screenshot 2023-06-30 203449

**Comment by papiasharmin** [comment]
> Spent a lot of time figuring it out but I only see the view all button which is not visible on volunteer account so that works but I can't see others. Maybe I am doing something wrong.
> 
> Screenshot 2023-06-30 203509 Screenshot 2023-06-30 203449

Hello @mahamakhan, thank you so much, I guess you might have just add team members and trying to view the button. some how the user team state is not updating, if you stop there server and rerun it with nmp run start:local, then log in to the account to see the button.

**Comment by Jiangwei-shi** [comment]
I login in as Administrator. 
image
In the dashboard. the “Leaderboard” view All button works as indeed.
image
However In the "Team Member Tasks", when I click the "View All" button several times, the screen freezes. I'm not sure if it's an issue with my computer. At least the first time I click the "View All" button, it works properly.
image

**Comment by xaanders** [comment]
I have tested the PR. Used admin, volunteer and core team accounts.

- Can see all the team members/tasks and leadership board with admin and core team: 

all

- Can see my teams as well as leadership board of my team:

teams

- Filter works good 
filter

- Volunteers and managers can't see these filters
vol
man


However if I remove myself from the team and assigned tasks, I still can see the filter buttons and teams I was before I removed myself(tried the same with core team and still could see the team I was removed from). 
After some time passed button disappeared from leadership board but remained on team member tasks(for core team user). I login to admin(without any team assigned) and back to core. 

removing

I assigned core team user to a project again. 
After 3-5 min button on leaderboard apeared(without login or out / restarting server). Team member tasks button appears after I log in admin account without any team assigned and then back to core. In this case both buttons show up.

**Review by xaanders — CHANGES_REQUESTED** [review]
I have tested the PR. Used admin, volunteer and core team accounts.

- Can see all the team members/tasks and leadership board with admin and core team: 

all

- Can see my teams as well as leadership board of my team:

teams

- Filter works good 
filter

- Volunteers and managers can't see these filters
vol
man


However if I remove myself from the team and assigned tasks, I still can see the filter buttons and teams I was before I removed myself(tried the same with core team and still could see the team I was removed from). 
After some time passed button disappeared from leadership board but remained on team member tasks(for core team user). I login to admin(without any team assigned) and back to core. 

removing

I assigned core team user to a project again. 
After 3-5 min button on leaderboard appeared(without login or out / restarting server). Team member tasks button appears after I log in admin account without any team assigned and then back to core. In this case both buttons show up.

**Review by one-community — CHANGES_REQUESTED** [review]
NEW to be fixed with this: New teams are not showing up now when I create a new user or try and find and add them to a team via the Profile Page. The only way I can add people to a Team is via the Teams page (Other Links --> Teams).

**Review by ramyaram2092 — COMMENTED** [review]
Hello @papiasharmin  is the work in progress for this PR?
I dont see the  View All /  My team button 

image

**Comment by papiasharmin** [comment]
> Hello @papiasharmin is the work in progress for this PR? I dont see the View All / My team button
> 
> image

Hello @ramyaram2092, if you have just assigned your self to a team, you won't be able to see that. I am working on it. I will try to finish this one by the end of this week and then let you know.

**Review by Changhao3220K — COMMENTED** [review]
hi @papiasharmin 
I am getting into issue that makes my local machine loading the tasks part all the time.
Screenshot 2023-11-17 at 2 33 28 PM
should I switch to this branch to check if the issue still presists?

**Review by Changhao3220K — COMMENTED** [review]
.

**Pull request closed without merging**